### PR TITLE
Remove bad tracks from 2-view BA

### PR DESCRIPTION
**Pull request by dellaert**
I enabled filtering after the bundle adjustment step, that filters the verified correspondences. 
I will do a comparison report after the CI finishes.

PR also includes 
- a test on triangulate_two_view_correspondences
- some re-formatting in `tests/common/test_gtsfm_data.py`, sorry about the noise.

### CODE REVIEW

File: gtsfm/common/gtsfm_data.py [hunk 1]
```diff
@@ -375,26 +375,29 @@ def __validate_track(self, track: SfmTrack, reproj_err_thresh: float) -> bool:
...
         # TODO: move this function to utils or GTSAM
         filtered_data = GtsfmData(self.number_images())
 
-        for j in range(self.number_tracks()):
-            track = self.get_track(j)
+        valid_mask = [self.__validate_track(track, reproj_err_thresh) for track in self._tracks]
 
-            if self.__validate_track(track, reproj_err_thresh):
+        for track, valid in zip(self._tracks, valid_mask):
+            if valid:
```
Comment (johnwlambert) [inline gtsfm/common/gtsfm_data.py:393]: prefer to `continue` early rather than nest logic here

File: gtsfm/two_view_estimator.py [hunk 2]
```diff
@@ -60,6 +60,7 @@ def __init__(
         bundle_adjust_2view: bool,
         eval_threshold_px: float,
         bundle_adjust_2view_maxiters: int = 100,
+        ba_reproj_error_thresh: float = 0.5,
```
Comment (ayushbaid) [inline gtsfm/two_view_estimator.py:63]: Isn't this too strict?

Comment (ayushbaid) [reply gtsfm/two_view_estimator.py:63]: For reference, we use 3px/4px threshold in verification and the final BA.

**Review by ayushbaid — APPROVED** [review]
LGTM. Please check the effect on performance as we are using a very strict threshold of 0.5px by default.

File: gtsfm/two_view_estimator.py (hunk 2 above)
Comment (dellaert) [reply gtsfm/two_view_estimator.py:63]: I'll let @johnwlambert comment, as he told me he fine-tuned these for his thesis. But this is 2-view bundle adjustment, which also optimizes for calibration, I think, so it should be able to overfit very closely. In reichstag it does not filter a lot of correspondences away.

**Comment by dellaert** [comment]
I ran a comparison and eager to hear your comments. Seems better overall, although several datasets did not run (time outs?). Some colors should maybe be reversed?
The entire file is [visual_comparison_dashboard.html.pdf](https://github.com/borglab/gtsfm/files/8647137/visual_comparison_dashboard.html.pdf) and the BA result:
image

**Comment by dellaert** [comment]
Update: failing CI was not a time out but a bug in a rare BA case, fixed, re-running.

**Comment by dellaert** [comment]
I now ran comparisons on 8 datasets, and also flipped some colors to better reflect what is "good" and what is "bad". Those edits are in another PR. I think everything is more positive, so I will merge after resolving conflicts, ok?

Full report: [visual_comparison_dashboard.html.pdf](https://github.com/borglab/gtsfm/files/8647867/visual_comparison_dashboard.html.pdf)

**Review by ayushbaid — APPROVED** [review]

**Comment by dellaert** [comment]
For posterity:
image

**Pull request merged**